src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,448 @@
+        vat.frob(ilk, address(pool), address(pool), address(pool), int256(amount), int256(amount));
+        // normalized debt == erc20 DAI (Vat rate for this ilk fixed to 1 RAY)
+        pool.deposit(amount);
+
+        // Verify the correct amount of gem shows up
+        uint256 newShares = pool.convertToShares(amount);
+        require(pool.shareBalance() >= _add(sharesPrev, newShares), "DssDirectDepositHub/no-receive-shares");
+
+        emit Wind(ilk, amount);
+    }
+
+    function _unwind(bytes32 ilk, D3MPoolLike pool, uint256 supplyReduction, uint256 availableLiquidity, Mode mode) internal {
+        // IMPORTANT: this function assumes Vat rate of this ilk will always be == 1 * RAY (no fees).
+        // That's why it converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+
+        uint256 assetBalance = pool.assetBalance();
+        uint256 daiDebt;
+        if (mode == Mode.NORMAL) {
+            // Normal mode or module just caged (no culled)
+            // debt is obtained from CDP art
+            (,daiDebt) = vat.urns(ilk, address(pool));
+        } else if (mode == Mode.MODULE_CULLED) {
+            // Module shutdown and culled
+            // debt is obtained from free collateral owned by this contract
+            daiDebt = vat.gem(ilk, address(pool));
+        } else {
+            // MCD caged
+            // debt is obtained from free collateral owned by the End module
+            end.skim(ilk, address(pool));
+            daiDebt = vat.gem(ilk, address(end));
+        }
+
+        // Unwind amount is limited by how much:
+        // - max reduction desired
+        // - liquidity available
+        // - gem we have to withdraw
+        // - dai debt tracked in vat (CDP or free)
+        uint256 amount = _min(
+                            _min(
+                                _min(
+                                    supplyReduction,
+                                    availableLiquidity
+                                ),
+                                assetBalance
+                            ),
+                            daiDebt
+                        );
+
+        // Determine the amount of fees to bring back
+        uint256 fees = 0;
+        if (assetBalance > daiDebt) {
+            fees = assetBalance - daiDebt;
+
+            if (_add(amount, fees) > availableLiquidity) {
+                // Don't need safe-math because this is constrained above
+                fees = availableLiquidity - amount;
+            }
+        }
+
+        if (amount == 0 && fees == 0) {
+            emit Unwind(ilk, 0);
+            return;
+        }
+
+        require(amount <= 2 ** 255, "DssDirectDepositHub/overflow");
+
+        // To save gas you can bring the fees back with the unwind
+        uint256 total = _add(amount, fees);
+        pool.withdraw(total);
```

Comment:
Same here, missing `erc20 => vat.dai` conversion.

Reply:
the withdraw and deposit functions in the TestPool contain `join`/`exit` calls to `daiJoin`

Reply:
https://github.com/makerdao/dss-direct-deposit/pull/31/commits/a6af7528bad745fccd7be1d460bc1f84591a3cb8

Reply:
Might it be a bit cleaner that the `join` happens to `address(this)` and then take the DAI in the `frob` from there as well. So it remains a bit more symmetric compared to the `_wind`.

Reply:
done https://github.com/makerdao/dss-direct-deposit/pull/31/commits/95f99aaee025a7b7a8ae1e2ca856c5127063a45a